app-ng/src/persistent-qvm.lisp

```diff
@@ -0,0 +1,223 @@
+;;;; persistent-qvm.lisp
+;;;;
+;;;; Author: appleby
+
+(in-package #:qvm-app-ng)
+
+(deftype persistent-qvm-token () 'string)
+
+(deftype persistent-qvm-state () '(member ready running waiting resuming dying))
+
+(defstruct (persistent-qvm (:constructor %make-persistent-qvm))
+  (qvm      (error "Must provide QVM")                                 :read-only t)
+  (cv       (error "Must provide CV")                                  :read-only t)
+  (lock     (error "Must provide LOCK")                                :read-only t)
+  (state    (error "Must provide STATE")    :type persistent-qvm-state)
+  (token    (error "Must provide TOKEN")    :type persistent-qvm-token :read-only t)
+  (metadata (error "Must provide METADATA") :type hash-table           :read-only t))
+
+(eval-when (:compile-toplevel :load-toplevel :execute)
+  (defun make-empty-persistent-qvms-db ()
+    (safety-hash:make-safety-hash :test 'equal)))
+
+(global-vars:define-global-var **persistent-qvms** (make-empty-persistent-qvms-db)
+  "The database of persistent QVMs. The keys are PERSISTENT-QVM-TOKENs and the values are PERSISTENT-QVMs.")
+
+(defun reset-persistent-qvms-db ()
+  "Reset the **PERSISTENT-QVMS** database."
+  (safety-hash:clrhash **persistent-qvms**))
+
+(defun persistent-qvms-count ()
+  "Return the number of PERSISTENT-QVMS currently allocated."
+  (safety-hash:hash-table-count **persistent-qvms**))
+
+(alexandria:define-constant +valid-pqvm-state-transitions+
+    '((ready    running           dying)
+      (running  ready    waiting  dying)
+      (waiting  resuming          dying)
+      (resuming running           dying)
+      (dying                      dying))
+  :test #'equal
+  :documentation "An alist of valid state transitions for a PERSISTENT-QVM. The alist is keyed on the current state. The value for each key is list of states that can be transitioned to from the corresponding current state.")
```

Comment:
An alist?

Reply:
I'm probably abusing the term here, but AFAIK an alist is just a list of conses where the car of each cons is a key and the cdr is an associated value. In this case, the value just happens to be a list.  I suppose I could reinforce this by spelling it `((foo . (bar baz)) ...)` rather than `((foo bar baz) ...)`

Reply:
Ah, yes! I see it now =)

Reply:
I can see how it would be confusing at a glance. I've added the `.`s to make the alist structure more apparent.